R/teal_data_utils.R

```diff
@@ -0,0 +1,69 @@
+#' `teal_data` utils
+#'
+#' In `teal` we need to recreate the `teal_data` object due to two operations:integer.max
+#' - we need to append filter-data code and objects which have been evaluated in `FilteredData` and
+#' we want to avoid double-evaluation.
+#' - we need to subset `teal_data` to datanames used by the module, to shorten obtainable R-code
+#'
+#' Due to above recreation of `teal_data` object can't be done simply by using public
+#' `teal.code` and `teal.data` methods.
+#'
+#' @param data (`teal_data`)
+#' @param code (`character`) code to append to `data@code`
+#' @param objects (`list`) objects to append to `data@env`
+#' @param datanames (`character`) names of the datasets
+#' @return modified `teal_data`
+#' @name teal_data_utilities
+NULL
+
+#' @rdname teal_data_utilities
+.append_evaluated_code <- function(data, code) {
+  checkmate::assert_class(data, "teal_data")
+  data@code <- c(data@code, code)
+  data@id <- c(data@id, sample.int(.Machine$integer.max, size = length(code)))
+  data@messages <- c(data@messages, rep("", length(code)))
+  data@warnings <- c(data@warnings, rep("", length(code)))
+  methods::validObject(data)
+  data
+}
+
+#' @rdname teal_data_utilities
+.append_modified_data <- function(data, objects) {
+  checkmate::assert_class(data, "teal_data")
+  checkmate::assert_class(objects, "list")
+  new_env <- list2env(objects, parent = .GlobalEnv)
+  rlang::env_coalesce(new_env, data@env)
+  data@env <- new_env
+  data
+}
+
+#' @rdname teal_data_utilities
+.subset_teal_data <- function(data, datanames) {
+  checkmate::assert_class(data, "teal_data")
+  checkmate::assert_class(datanames, "character")
+  new_data <- data
+  datasets_to_extract <- intersect(c(datanames, sprintf("%s_raw", datanames)), ls(data@env))
+  new_data@code <- get_code_dependency(data@code, datasets_to_extract)
+  new_data@id <- sample.int(.Machine$integer.max, size = length(new_data@code))
+  new_data@warnings <- rep("", length(new_data@code))
+  new_data@messages <- rep("", length(new_data@code))
```

Comment:
What if `data` object had anything in `@warnings` or `@message` already? Do they get over-written by `""`?

Reply:
This is a difficult task - because we don't have indices of the `@code` elements extracted with `get_code_dependencies`. Ignored @warnings was also the case before this change

Reply:
Maybe we can extend `get_code_dependencies` so it returns indices? Not necessarily during this refactor, but once when we have more time?